### PR TITLE
snap_restore: added retry for recovery

### LINKED ISSUE #46181: ebs br: support retry to restore at recoverregion and waitapply phase
## Enhancement
Referencing https://github.com/pingcap/tidb-operator/issues/5236

### PR DESCRIPTION
**Pull request by YuJuncen**
### What problem does this PR solve?

Issue Number: close #46181

Problem Summary:
Now, when TiKV reboots, it is possible to abort the whole restore procedure.

This PR has also introduced a new interface for retrying, `WithRetryV2`, which supports returned value and move the closure to be retried last of the arguments. (So the code will probably be more readable, comparing `WithRetry(ctx, backoffer, func() { /* multi lines */ })` and `WithRetry(ctx, func() { /* multi lines */ }, backoffer)`.

### What is changed and how it works?
This PR added retry to the `RecoverData` call.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Make snapshot restoration more robust.
```

### CODE REVIEW

**Review by BornChanger — APPROVED** [review]

File: br/pkg/task/restore_data.go [hunk 1]
```diff
@@ -146,6 +146,8 @@ func RunResolveKvData(c context.Context, g glue.Glue, cmdName string, cfg *Resto
 	// restore tikv data from a snapshot volume
 	var totalRegions int
 
+	// Roughly handle the case that some TiKVs are rebooted during making plan.
+	// Generally,
```
Comment (Leavrth) [inline br/pkg/task/restore_data.go:150]: not entire annotation?

File: br/pkg/restore/data.go [hunk 2]
```diff
@@ -35,39 +104,50 @@ import (
...
 	if err := recovery.ReadRegionMeta(ctx); err != nil {
-		return 0, errors.Trace(err)
+		return 0, recoveryError{error: err, atStage: StageCollectingMeta}
 	}
 
 	totalRegions := recovery.GetTotalRegions()
 
-	if err := recovery.MakeRecoveryPlan(); err != nil {
-		return totalRegions, errors.Trace(err)
+	err := recovery.MakeRecoveryPlan()
```
Comment (Leavrth) [inline br/pkg/restore/data.go:124]: keep the format as before?, such as
```
if err := xxx; err != nil {
```

File: br/pkg/restore/data.go [hunk 3]
```diff
@@ -35,39 +104,50 @@ import (
...
 	}
 
 	if err := recovery.WaitApply(ctx); err != nil {
-		return totalRegions, errors.Trace(err)
+		return totalRegions, recoveryError{error: err, atStage: StageRecovering}
 	}
-
-	if err := recovery.PrepareFlashbackToVersion(ctx, resolveTS, restoreTS-1); err != nil {
-		return totalRegions, errors.Trace(err)
+	err = recovery.PrepareFlashbackToVersion(ctx, resolveTS, restoreTS-1)
```
Comment (Leavrth) [inline br/pkg/restore/data.go:144]: ditto

File: br/pkg/restore/data.go [hunk 4]
```diff
@@ -27,6 +27,75 @@ import (
...
+	return StageUnknown
+}
+
+type recoveryBackoffer struct {
+	state utils.RetryState
+}
+
+func newRecoveryBackoffer() *recoveryBackoffer {
+	return &recoveryBackoffer{
+		state: utils.InitialRetryState(16, 0, 0),
```
Comment (Leavrth) [inline br/pkg/restore/data.go:76]: is the parameter `0` ok?

**Review by Leavrth — APPROVED** [review]

**Pull request merged**